EntityGenerator lifecycle callbacks

timum-viw opened the pull request:
This PR addresses a problem in the Entity Generator. when adding the same lifecycle event callback to two or more lifecycle events the generator will create a stub for each event each with the same method name. This results in fatal 'Cannot redeclare' errors. This Problem occurred only if the callback name contains uppercase letters.

this PR intends to fix this issue by updating the EntityGenerator to allow the same callback to be triggered by multiple event.

timum-viw committed:
add breaking test cases
When adding the same lifecycle event callback to two or more lifecycle events, the generator will create a stub for each event resulting in fatal 'Cannot redeclare' errors. That is, only if the callback name contains uppercase letters.

timum-viw committed:
fix generating duplicate method stubs
this will fix the fatal errors when having the same lifecycle event callback on multiple lifecycle events. Tests will still fail, since the necessary annotations for multiple lifecycle events are not properly added.

timum-viw committed:
fix annotations for lifecycle callbacks
this will fix adding annotations for all referenced lifecycle events when creating lifecycle callback stubs. It takes case insensitivity into account.

greg0ire commented:
It seems there are CI jobs failing. Please take a look at [this guide](https://www.doctrine-project.org/contribute/#dealing-with-checks-and-tools) for more on how to handle those.

timum-viw committed:
fix formatting

timum-viw commented:
Sorry about that. I fixed the formatting according to phpcs output and locally no errors in the changed files are reported.

greg0ire commented on src/Tools/EntityGenerator.php:
```diff
@@ -1408,21 +1421,26 @@ protected function generateEntityStubMethod(ClassMetadataInfo $metadata, $type,
      * @return string
      */
-    protected function generateLifecycleCallbackMethod($name, $methodName, ClassMetadataInfo $metadata)
+    protected function generateLifecycleCallbackMethod(array $events, string $methodName, ClassMetadataInfo $metadata)
```
This is a breaking change for extending classes. Please revert this part and resort to detecting the type of `$name` to decide what to do inside the method.

timum-viw replied:
Will do. Thank for the guidance.

timum-viw committed:
revert parameter type change
fallback to parameter type detection in method body to not break extending classes

greg0ire commented:
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/2.20.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

greg0ire commented:
Argh, wrong branch… well, you will have to wait for 2.20.0 to get your bugfix. Thanks for working on that anyway, and congrats on your first contribution.

timum-viw commented:
Yeah, noticed that too when I read the part in the docs about picking the right branch 😆 thanks for your fast replies and your general work on doctrine!
